File: generated_codes/chat_agents/chat-agents-java/skiplist/claude_java/output7_skiplist.java
package com.example.Sets;

import com.example.utils.SetADT;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class ConcurrentDataStructure implements SetADT {

    private static final int MAX_LEVEL = 16;
    private static final double PROBABILITY = 0.5;

    private static class Node {
        final int key;
        final int level;
        final AtomicMarkableReference<Node>[] forward;

        @SuppressWarnings("unchecked")
        Node(int key, int level) {
            this.key = key;
            this.level = level;
            this.forward = new AtomicMarkableReference[level + 1];
            for (int i = 0; i <= level; i++) {
                this.forward[i] = new AtomicMarkableReference<>(null, false);
            }
        }
    }

    private final Node head;
    private final Node tail;
    private final java.util.concurrent.atomic.AtomicInteger randomSeed;

    @SuppressWarnings("unchecked")
    public ConcurrentDataStructure() {
        head = new Node(Integer.MIN_VALUE, MAX_LEVEL);
        tail = new Node(Integer.MAX_VALUE, MAX_LEVEL);
        for (int i = 0; i <= MAX_LEVEL; i++) {
            head.forward[i] = new AtomicMarkableReference<>(tail, false);
        }
        randomSeed = new java.util.concurrent.atomic.AtomicInteger(0x12345678);
    }

    private int randomLevel() {
        int level = 0;
        int seed;
        int next;
        do {
            seed = randomSeed.get();
            next = seed * 1664525 + 1013904223;
        } while (!randomSeed.compareAndSet(seed, next));
        int bits = next;
        while ((bits & 1) == 1 && level < MAX_LEVEL) {
            level++;
            bits >>= 1;
        }
        return level;
    }

    @SuppressWarnings("unchecked")
    private Node[] find(int key, Node[] preds, Node[] succs) {
        boolean[] marked = {false};
        boolean snip;
        Node pred, curr, succ;

        retry:
        while (true) {
            pred = head;
            for (int level = MAX_LEVEL; level >= 0; level--) {
                curr = pred.forward[level].getReference();
                while (true) {
                    succ = curr.forward[level].get(marked);
                    while (marked[0]) {
                        snip = pred.forward[level].compareAndSet(curr, succ, false, false);
                        if (!snip) continue retry;
                        curr = pred.forward[level].getReference();
                        succ = curr.forward[level].get(marked);
                    }
                    if (curr.key < key) {
                        pred = curr;
                        curr = succ;
                    } else {
                        break;
                    }
                }
                preds[level] = pred;
                succs[level] = curr;
            }
            return succs;
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public boolean add(int key) {
        int topLevel = randomLevel();
        Node[] preds = new Node[MAX_LEVEL + 1];
        Node[] succs = new Node[MAX_LEVEL + 1];

        while (true) {
            find(key, preds, succs);
            if (succs[0] != null && succs[0].key == key) {
                return false;
            }
            Node newNode = new Node(key, topLevel);
            for (int level = 0; level <= topLevel; level++) {
                newNode.forward[level] = new AtomicMarkableReference<>(succs[level], false);
            }
            Node pred = preds[0];
            Node succ = succs[0];
            if (!pred.forward[0].compareAndSet(succ, newNode, false, false)) {
                continue;
            }
            for (int level = 1; level <= topLevel; level++) {
                while (true) {
                    pred = preds[level];
                    succ = succs[level];
                    if (pred.forward[level].compareAndSet(succ, newNode, false, false)) {
                        break;
                    }
                    find(key, preds, succs);
                }
            }
            return true;
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public boolean remove(int key) {
        Node[] preds = new Node[MAX_LEVEL + 1];
        Node[] succs = new Node[MAX_LEVEL + 1];
        boolean[] marked = {false};

        while (true) {
            find(key, preds, succs);
            Node target = succs[0];
            if (target == null || target.key != key) {
                return false;
            }
            int topLevel = target.level;
            for (int level = topLevel; level >= 1; level--) {
                Node succ = target.forward[level].get(marked);
                while (!marked[0]) {
                    target.forward[level].compareAndSet(succ, succ, false, true);
                    succ = target.forward[level].get(marked);
                }
            }
            Node succ = target.forward[0].get(marked);
            while (true) {
                boolean markedIt = target.forward[0].compareAndSet(succ, succ, false, true);
                if (markedIt) {
                    // Node has been marked
                    find(key, preds, succs);
                    return true;
                }
                succ = target.forward[0].get(marked);
                if (marked[0]) {
                    return false;
                }
            }
        }
    }

    @Override
    public boolean contains(int key) {
        boolean[] marked = {false};
        Node curr = head;
        for (int level = MAX_LEVEL; level >= 0; level--) {
            Node next = curr.forward[level].getReference();
            while (next.key < key) {
                curr = next;
                next = curr.forward[level].getReference();
            }
        }
        curr = curr.forward[0].getReference();
        curr.forward[0].get(marked);
        return curr.key == key && !marked[0];
    }
}

File: generated_codes/chat_agents/chat-agents-cpp/skiplist/gemini/output7_skiplist.cpp
#pragma once
#include "../utils/SetADT.hpp"
#include <atomic>
#include <climits>
#include <cstdint>
#include <random>

class ConcurrentDataStructure : public SetADT {
private:
    static constexpr int MAX_LEVEL = 16;

    struct Node {
        int val;
        int topLevel;
        std::atomic<Node*> forward[MAX_LEVEL];

        Node(int key, int height) : val(key), topLevel(height) {
            for (int i = 0; i < MAX_LEVEL; ++i) {
                forward[i].store(nullptr, std::memory_order_relaxed);
            }
        }
    };

    Node* head;
    Node* tail;

    static bool is_marked_ref(Node* p) {
        return (reinterpret_cast<uintptr_t>(p) & 1L) != 0;
    }

    static Node* get_unmarked_ref(Node* p) {
        return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) & ~1L);
    }

    static Node* get_marked_ref(Node* p) {
        return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) | 1L);
    }

    int random_level() {
        static thread_local std::mt19937 generator(std::random_device{}());
        static thread_local std::geometric_distribution<int> distribution(0.5);
        int lvl = distribution(generator);
        if (lvl >= MAX_LEVEL) return MAX_LEVEL - 1;
        return lvl;
    }

    bool find(int key, Node** preds, Node** succs) {
        int retry_count = 0;
    retry:
        Node* pred = head;
        for (int level = MAX_LEVEL - 1; level >= 0; --level) {
            Node* curr = pred->forward[level].load(std::memory_order_acquire);
            while (true) {
                Node* succ = curr;
                while (is_marked_ref(succ)) {
                    curr = get_unmarked_ref(succ);
                    succ = curr->forward[level].load(std::memory_order_acquire);
                }
                
                Node* unmarked_curr = get_unmarked_ref(curr);
                if (unmarked_curr->val >= key) {
                    preds[level] = pred;
                    succs[level] = unmarked_curr;
                    break;
                }
                
                Node* expected = curr;
                if (!pred->forward[level].compare_exchange_weak(expected, unmarked_curr, std::memory_order_acq_rel, std::memory_order_acquire)) {
                    if (++retry_count > 2) {
                        retry_count = 0;
                        goto retry;
                    }
                    goto retry;
                }
                pred = unmarked_curr;
                curr = pred->forward[level].load(std::memory_order_acquire);
            }
        }
        return (succs[0]->val == key);
    }

public:
    ConcurrentDataStructure() {
        head = new Node(INT_MIN, MAX_LEVEL);
        tail = new Node(INT_MAX, MAX_LEVEL);
        for (int i = 0; i < MAX_LEVEL; ++i) {
            head->forward[i].store(tail, std::memory_order_relaxed);
        }
    }

    ~ConcurrentDataStructure() override {
        Node* curr = head;
        while (curr != nullptr) {
            Node* next = get_unmarked_ref(curr->forward[0].load(std::memory_order_relaxed));
            delete curr;
            curr = next;
        }
    }

    bool contains(int key) override {
        Node* pred = head;
        for (int level = MAX_LEVEL - 1; level >= 0; --level) {
            Node* curr = pred->forward[level].load(std::memory_order_acquire);
            while (true) {
                Node* succ = curr;
                while (is_marked_ref(succ)) {
                    curr = get_unmarked_ref(succ);
                    succ = curr->forward[level].load(std::memory_order_acquire);
                }
                Node* unmarked_curr = get_unmarked_ref(curr);
                if (unmarked_curr->val >= key) {
                    if (unmarked_curr->val == key) {
                        Node* next_level_ref = unmarked_curr->forward[0].load(std::memory_order_acquire);
                        return !is_marked_ref(next_level_ref);
                    }
                    break;
                }
                pred = unmarked_curr;
                curr = pred->forward[level].load(std::memory_order_acquire);
            }
        }
        return false;
    }

    bool add(int key) override {
        int topLevel = random_level();
        Node* preds[MAX_LEVEL];
        Node* succs[MAX_LEVEL];

        while (true) {
            if (find(key, preds, succs)) {
                Node* found_node = succs[0];
                Node* next_ref = found_node->forward[0].load(std::memory_order_acquire);
                if (!is_marked_ref(next_ref)) {
                    return false;
                }
                continue;
            }

            Node* newNode = new Node(key, topLevel + 1);
            for (int i = 0; i <= topLevel; ++i) {
                newNode->forward[i].store(succs[i], std::memory_order_relaxed);
            }

            Node* pred = preds[0];
            Node* succ = succs[0];
            if (pred->forward[0].compare_exchange_strong(succ, newNode, std::memory_order_acq_rel, std::memory_order_acquire)) {
                for (int i = 1; i <= topLevel; ++i) {
                    while (true) {
                        pred = preds[i];
                        succ = succs[i];
                        newNode->forward[i].store(succ, std::memory_order_relaxed);
                        if (pred->forward[i].compare_exchange_strong(succ, newNode, std::memory_order_acq_rel, std::memory_order_acquire)) {
                            break;
                        }
                        find(key, preds, succs);
                    }
                }
                return true;
            }
            delete newNode;
        }
    }

    bool remove(int key) override {
        Node* preds[MAX_LEVEL];
        Node* succs[MAX_LEVEL];

        while (true) {
            if (!find(key, preds, succs)) {
                return false;
            }

            Node* nodeToRemove = succs[0];
            bool marked_by_me = false;

            for (int level = nodeToRemove->topLevel - 1; level >= 0; --level) {
                while (true) {
                    Node* succ = nodeToRemove->forward[level].load(std::memory_order_acquire);
                    if (is_marked_ref(succ)) {
                        break;
                    }
                    Node* marked_succ = get_marked_ref(succ);
                    if (nodeToRemove->forward[level].compare_exchange_strong(succ, marked_succ, std::memory_order_acq_rel, std::memory_order_acquire)) {
                        if (level == 0) {
                            marked_by_me = true;
                        }
                        break;
                    }
                }
            }

            if (marked_by_me) {
                find(key, preds, succs);
                return true;
            } else {
                Node* check_ref = nodeToRemove->forward[0].load(std::memory_order_acquire);
                if (is_marked_ref(check_ref)) {
                    return false;
                }
            }
        }
    }
};
